### DISCUSSION TITLE
Intel SNB/IVB/HSW/BDW : IMC extended timings quest

### DISCUSSION THREAD

**Discussion by cyring** [cyring/CoreFreq#342] [General]
# Reverse protocol
## Specification
Missing specs from `0x4008` to `0x4024`

## Dump
```
Channel[0]
0x4000 [001b899a]
0x4004 [90248588]
0x4008 [0a022220]
0x400c [000058d7]
0x4010 [00000000]
0x4014 [00000000]
0x4018 [00000000]
0x401c [00100000]
0x4020 [00100005]
0x4024 [27272020]
0x4028 [000e2200]
0x402c [00000000]
Channel[1]
0x4000 [001b899a]
0x4004 [90248588]
0x4008 [0a022220]
0x400c [000058d7]
0x4010 [00000000]
0x4014 [00000000]
0x4018 [00000000]
0x401c [00100000]
0x4020 [00100005]
0x4024 [28282020]
0x4028 [000e3300]
0x402c [00000000]
```
## BIOS

![DSC_0072](https://user-images.githubusercontent.com/11563789/169660509-0aad1c06-6419-47ca-82a7-1a363113b899.JPG)
![DSC_0073](https://user-images.githubusercontent.com/11563789/169660510-fb1fdee8-25e5-48a4-ab26-8c6259567133.JPG)
![DSC_0074](https://user-images.githubusercontent.com/11563789/169660513-0979c89b-0a04-49dd-82ef-21b61e4f4e6c.JPG)
![DSC_0075](https://user-images.githubusercontent.com/11563789/169660514-d82b3ca9-2683-42e1-b275-1b9a787d0f63.JPG)

## Source
```c
void Query_SNB_IMC(void __iomem *mchmap, unsigned short mc)
{
           /* ... */
    for (cha = 0; cha < PUBLIC(RO(Proc))->Uncore.MC[mc].ChannelCount; cha++)
    {
        unsigned int    R4000=0, R4004=0, R4008=0, R400c=0, R4010=0, R4014=0,
                        R4018=0, R401c=0, R4020=0, R4024=0, R4028=0, R402c=0;

                R4000 = readl(mchmap + 0x4000 + 0x400 * cha);
                R4004 = readl(mchmap + 0x4004 + 0x400 * cha);
                R4008 = readl(mchmap + 0x4008 + 0x400 * cha);
                R400c = readl(mchmap + 0x400c + 0x400 * cha);
                R4010 = readl(mchmap + 0x4010 + 0x400 * cha);
                R4014 = readl(mchmap + 0x4014 + 0x400 * cha);
                R4018 = readl(mchmap + 0x4018 + 0x400 * cha);
                R401c = readl(mchmap + 0x401c + 0x400 * cha);
                R4020 = readl(mchmap + 0x4020 + 0x400 * cha);
                R4024 = readl(mchmap + 0x4024 + 0x400 * cha);
                R4028 = readl(mchmap + 0x4028 + 0x400 * cha);
                R402c = readl(mchmap + 0x402c + 0x400 * cha);

                printk( "Channel[%d]\n"         \
                        "0x4000 [%08x]\n"       \
                        "0x4004 [%08x]\n"       \
                        "0x4008 [%08x]\n"       \
                        "0x400c [%08x]\n"       \
                        "0x4010 [%08x]\n"       \
                        "0x4014 [%08x]\n"       \
                        "0x4018 [%08x]\n"       \
                        "0x401c [%08x]\n"       \
                        "0x4020 [%08x]\n"       \
                        "0x4024 [%08x]\n"       \
                        "0x4028 [%08x]\n"       \
                        "0x402c [%08x]\n",
                        cha,
                        R4000,
                        R4004,
                        R4008,
                        R400c,
                        R4010,
                        R4014,
                        R4018,
                        R401c,
                        R4020,
                        R4024,
                        R4028,
                        R402c);
    }
           /* ... */
}

**Comment by cyring** [comment]
# Guess round `#1`
## BIOS (Third Timings)
![DSC_0001](https://user-images.githubusercontent.com/11563789/169667479-f7216883-aca6-42e7-8c13-5220ec0cdad9.JPG)

## Dump
```
                              Channel[0]
                              0x4000 [001b899a]
                              0x4004 [90248588]
                              0x4008 [cc025761]
                              0x400c [000058d7]
                              0x4010 [00000000]
                              0x4014 [00000000]
                              0x4018 [00000000]
                              0x401c [00100000]
                              0x4020 [00100005]
                              0x4024 [27272020]
                              0x4028 [000e2200]
                              0x402c [00000000]
                              Channel[1]
                              0x4000 [001b899a]
                              0x4004 [90248588]
                              0x4008 [cc025761]
                              0x400c [000058d7]
                              0x4010 [00000000]
                              0x4014 [00000000]
                              0x4018 [00000000]
                              0x401c [00100000]
                              0x4020 [00100005]
                              0x4024 [28282020]
                              0x4028 [000e3300]
                              0x402c [00000000]
```

## _CoreFreq_
![2022-05-21-232007_643x430_scrot](https://user-images.githubusercontent.com/11563789/169669390-2715c6c2-0aa1-43be-a9cb-05e1147b824b.png)

**Comment by cyring** [comment]
# Guess round `#2`
## BIOS `tRRSR`

![DSC_0002](https://user-images.githubusercontent.com/11563789/169670012-eda6a4b8-cec5-4d40-a7c8-af9e26f01e29.JPG)

## Dump
```
                              Channel[0]
                              0x4000 [001b899a]
                              0x4004 [90248588]
                              0x4008 [1a022220]
                              0x400c [000058d7]
                              0x4010 [00000000]
                              0x4014 [00000000]
                              0x4018 [00000000]
                              0x401c [00100000]
                              0x4020 [00100005]
                              0x4024 [27272020]
                              0x4028 [000e2200]
                              0x402c [00000000]
                              Channel[1]
                              0x4000 [001b899a]
                              0x4004 [90248588]
                              0x4008 [1a022220]
                              0x400c [000058d7]
                              0x4010 [00000000]
                              0x4014 [00000000]
                              0x4018 [00000000]
                              0x401c [00100000]
                              0x4020 [00100005]
                              0x4024 [28282020]
                              0x4028 [000e3300]
                              0x402c [00000000]
```

## _CoreFreq_

![2022-05-22-001820_644x434_scrot](https://user-images.githubusercontent.com/11563789/169670727-698cf953-d6d3-4e77-ba5b-eec612029d3c.png)

**Comment by cyring** [comment]
# Guess round `#3`
## BIOS

![DSC_0003](https://user-images.githubusercontent.com/11563789/169671176-8cb3add1-42eb-4d3d-8d93-ad5755fc6106.JPG)

## Dump
```
                              Channel[0]
                              0x4000 [001b899a]
                              0x4004 [90248588]
                              0x4008 [0a022220]
                              0x400c [000058d7]
                              0x4010 [00000000]
                              0x4014 [00000000]
                              0x4018 [00000000]
                              0x401c [00100000]
                              0x4020 [00100005]
                              0x4024 [27272020]
                              0x4028 [000e2200]
                              0x402c [00000000]
                              Channel[1]
                              0x4000 [001b899a]
                              0x4004 [90248588]
                              0x4008 [0a022220]
                              0x400c [000058d7]
                              0x4010 [00000000]
                              0x4014 [00000000]
                              0x4018 [00000000]
                              0x401c [00100000]
                              0x4020 [00100005]
                              0x4024 [28282020]
                              0x4028 [000e3300]
                              0x402c [00000000]
```

**Comment by svmlegacy** [comment]
@cyring 
Here's a new gist with the outputs from my Sandy Bridge Celeron G550.

[Here.](https://gist.github.com/svmlegacy/23344390dded8687e07d9d4614bc4a1d)

(This is a response to this [note](https://gist.github.com/svmlegacy/0b799f2842693ad396486afb06960bb3?permalink_comment_id=4179828#gistcomment-4179828)) This is the same system, but I've been slowing working up the CPU performance...

**Reply by cyring** [reply]
Your outputs are really helpful. Thank you.

So what we have more now is `B2B` and most third timings.
To compare with, does this platform has a BIOS showing them ?

**Comment by cyring** [comment]
Need someone to confirm with screenshot the IMC result on Haswell/Broadwell.
Especially platform wirh host bridge of DID  `0x0c00` or `0x0c04`
Please pull latest commits from the `develop` branch; rebuild, reload _CoreFreq_ and post results here.

Focus on MC timings and DIMM geometry.
Recently SNB, IVB and Atom are progressively being added. Feel free to show me your results.

EDIT: I also need those Xeon's architectures.
Please give a try and show me. But not on Production machine or during down time you can boot my all-in-one developer ISO with hybrid mbr/uefi ArchLinux and _CoreFreq_ embedded. 
Disks are not mounted; ramfs only.
Go to www.cyring.fr to download ISO.

**Comment by svmlegacy** [comment]
[Haswell-DT Intel Core i7-4570 SR14E](https://gist.github.com/svmlegacy/729c832d38d879551f8af2050f044a36)
[Broadwell-H Intel Xeon E3-1270L v4 SR2AZ](https://gist.github.com/svmlegacy/0ee61bf007c4ca2b3d7c90300cb9f99c)

Bios pictures will go into the comments of each respectively. (Along with the G550)

**Reply by cyring** [reply]
Awesome!

I will come back with an attempt to decode the Broadwell Xeon IMC.

And minor fixes among both processors.

**Comment by svmlegacy** [comment]
[Sandy Bridge-EP Intel Xeon E5-2640 SR0KR](https://gist.github.com/svmlegacy/4f6de77103dce916d57f05016f2b6243)
[Ivy Bridge-EP Intel Xeon E5-1620 v2 SR1AR](https://gist.github.com/svmlegacy/7a2ba9233c903b2a5589726d0d7a98a7)

**Comment by cyring** [comment]
@svmlegacy Hello

For IMC testings is available the commit 347dd79364c7fff585f931e98eaec73e0d45ab6d for the Broadwell E3-1270L and the Xeon E5-2640 processors.

**Reply by svmlegacy** [reply]
Both gists have a new file for https://github.com/cyring/CoreFreq/commit/347dd79364c7fff585f931e98eaec73e0d45ab6d.

Broadwell-H - Seems to pulling an incorrect memory speed. It is running at 1333 MHz (And max XMP is 2133)
Sandy Bridge-EP - Looks like it's getting good numbers for 1st and 2nd channel. 3 & 4 don't seem to be pulling timings correctly.

**Reply by cyring** [reply]
> Both gists have a new file for [347dd79](https://github.com/cyring/CoreFreq/commit/347dd79364c7fff585f931e98eaec73e0d45ab6d).
> 
> Broadwell-H - Seems to pulling an incorrect memory speed. It is running at 1333 MHz (And max XMP is 2133) Sandy Bridge-EP - Looks like it's getting good numbers for 1st and 2nd channel. 3 & 4 don't seem to be pulling timings correctly.

SNB-EP IMC channels and DIMMs counting is refactored. Available in last commit.
For your testings, just the IMC output will be enough.

I'm also introducing code for Haswell-EP in case you have such Xeon ?

Btw, is this [X9DRD](https://www.supermicro.com/wdl/ISO_Extracted/CDR-X9_1.30_for_Intel_X9_platform/MANUALS/X9DRD-iF_LF.pdf) your board datasheet ?

**Reply by svmlegacy** [reply]
![Screenshot from 2022-05-30 20-35-34](https://user-images.githubusercontent.com/59940364/171071595-9276928c-dfd0-4337-a23d-38d2af3b3d49.png)

Commit gcd4cdfb

And Correct datasheet. This board is the Supermicro X9DRD-LF. I don't have a 2nd cooler for it, but I can do some short testings in 2P operation if you like.

**Reply by svmlegacy** [reply]
I should be able to test Haswell-E sometime soon. I don't have any Haswell-EP or Broadwell-EP at the moment.

**Reply by cyring** [reply]
**Thank you** for your tests.

* DIMM count is improved but channels count is still an issue.
  Do you confirm the platform is quad channel and only one DIMM is populated every two slots ?

* Yes I would like a test with your `Haswell-E`
  Not sure where it will fall through. Here all the Haswell I have set in table:
https://github.com/cyring/CoreFreq/blob/cd4cdfbb1bf0c061a7e73dd087dfc55e61324395/corefreqk.h#L1771
  The IMC decoding is following the PCI ID. So result could be the same as `EP`
  If no IMC is showing up then I miss those IDs. `lspci -nn` of the `Haswell-E` will then be useful to me.

* Dual sockets testings: I will appreciate the complete overview from the CLI; especially the Core topology.
  Core & socket Temperature, Core voltage, Power consumed are also of a great interest. 
  I mean If _CoreFreq_ is able to monitor  discrete values among both sockets ?
  Suggest to change their scope in the Settings menu if they don't show up in the UI.
  If possible w/o a cooler, you will have to stress socket differently to catch any sensor difference among both.

**Reply by cyring** [reply]
@svmlegacy 
There's a new commit for QPI frequency of Xeon
 BUS Rate of IMC should display 7200

Another thought is that you have a single processor socket populated, perhaps quad channels timings are set for its connected DIMMs only ?

**Reply by svmlegacy** [reply]
![image](https://user-images.githubusercontent.com/59940364/171519268-c3f5739a-9779-4274-a145-1284e75e2b59.png)

Sandy-E and Ivy-E are quad channel CPU's, so in 2P operation I'd actually expect to see eight total channels. The current config has all 4 DIMM's connected to the first CPU only.

**Reply by svmlegacy** [reply]
[2x Intel Xeon E5-2640 SR0KR](https://gist.github.com/svmlegacy/734b7b123d7699f34dd2dd5c42876f3f)

DIMMs as follows:
CPU 0: B1, C1
CPU 1: E1, G1

Odd config as I need to move the dimms to clear coolers... 2 memory channels per CPU in this config.

**Reply by cyring** [reply]
Oh, a topology problem I have already encountered within previous architecture use cases:
* Two controllers assigned to the same DID. 
  Bellow `3ca8` will be probbed twice as controller `#0`, rather than controllers `#0` then `#1`
```
7f:0f.0 System peripheral [0880]: Intel Corporation Xeon E5/Core i7 Integrated Memory Controller Registers [8086:3ca8] (rev 07)
```
```
ff:0f.0 System peripheral [0880]: Intel Corporation Xeon E5/Core i7 Integrated Memory Controller Registers [8086:3ca8] (rev 07)
```
* So I have to refactor the topology discovering; not relying on that kernel DID enumeration; and do a direct BDF addressing per processor socket. Similar to what I did for `Jasper`
https://github.com/cyring/CoreFreq/blob/c9cb78ff504d500fdfdc19b23fd3b6dbc960865a/corefreqk.c#L4895

**Comment by svmlegacy** [comment]
[Haswell-E Intel Core i7 5960X SR20Q](https://gist.github.com/svmlegacy/f9578b7c70b4100e9add3d9d0f7b4e26)
(Only two memory channels populated.)

**Reply by cyring** [reply]
Hello,

Can you pull latest _CoreFreq_ and print the output of this Haswell-E ?

**Reply by svmlegacy** [reply]
Of course. Here's a screenshot of the latest CoreFreq Develop:
![image](https://user-images.githubusercontent.com/59940364/173254691-5719ff8f-f852-442c-a0ac-7d705c074347.png)

**Reply by cyring** [reply]
If total RAM is still 16GB then there"s a regression in DIMM topology

**Reply by svmlegacy** [reply]
@cyring No regression, this motherboard is very picky on POST. Could only get it to light up with a single DIMM.

1x Corsair 8 GB DIMM in channel #0.

**Reply by cyring** [reply]
This might explain why I can not read DDR3 speed.

Do you have any other Haswell Destop and/or mobile, not Xeon, to confirm IMC decoding ?

**Reply by cyring** [reply]
This [i5-4570](https://gist.github.com/svmlegacy/729c832d38d879551f8af2050f044a36) for IMC again, please.

**Reply by svmlegacy** [reply]
![image](https://user-images.githubusercontent.com/59940364/173259412-e6fdab7a-6c00-4342-bcae-5e4fe6e6907e.png)
Haswell-DT, which is DDR3. Remember that Haswell-E is DDR4 (And I think Haswell-E/EP has DDR3 onboard like Skylake, but I can't test that...), so these are probably very different memory controllers.

**Reply by cyring** [reply]
# `i5-4570`
## Timings
Setting | _CoreFreq_ | BIOS
----------|------------------|--------
`tRRD` | `5` | `4`
`tWR` | N/A  | `10`
`tFAW` | N/A  | `20`
`tRTP` | N/A  | `5`
`tWTP` | N/A  | `21`
`drWR` | `4` | `5`
`tCKE` | N/A | `4`

## Geometry
```
( 8 x Bank x Rank x Rows x Cols ) / ( 1024 x 1024 )
```
```
( 8 x 8 x 2 x 16384 x 1024 ) / ( 1024 x 1024) = 2048 MB vs 8192 MB
```

**Comment by cyring** [comment]
@svmlegacy Hello

For the 2x Intel Xeon E5-2640 setup, can you please run the attached archive.
In the Kernel log should be printed some PCI CPUBUSNO register(s) as bellow:
```
CoreFreq: CPUBUSNO[0x00000000] @ BDF(0x0,0x0,0x108)
```
[CoreFreq_develop.tar.gz](https://github.com/cyring/CoreFreq/files/8839080/CoreFreq_develop.tar.gz)

**Reply by svmlegacy** [reply]
As follows:

```
CoreFreq(0:12:-1): Processor [ 06_2D] Architecture [SandyBridge/EP/Romley] SMT [24/24]
CoreFreq: CPUBUSNO[0x17f00] @ BDF(0x0,0x3c28,0x108)
```

**Reply by cyring** [reply]
> As follows:
> 
> ```
> CoreFreq(0:12:-1): Processor [ 06_2D] Architecture [SandyBridge/EP/Romley] SMT [24/24]
> CoreFreq: CPUBUSNO[0x17f00] @ BDF(0x0,0x3c28,0x108)
> ```

Thank you.

Based on E5 [datasheet](https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/xeon-e5-1600-2600-vol-2-datasheet.pdf), these are the two IMC I'm trying to map:

![2022-06-06-082718_787x615_scrot](https://user-images.githubusercontent.com/11563789/172107792-71db0702-17fe-44d7-b410-c8c65050dbbd.png)


`CPUBUSNO` is specified as:

![2022-06-06-083227_788x168_scrot](https://user-images.githubusercontent.com/11563789/172108635-6d76b109-07d2-4b9c-bab5-a4060a2d3e16.png)
![2022-06-06-083242_779x360_scrot](https://user-images.githubusercontent.com/11563789/172108645-cc770512-eef2-4b64-96ed-bf75dfd56726.png)

With a dump value of `0x17f00` we have:
* IIO/PCH bus number `0x00`
* Uncore bus number `0x7f`
* Valid bit is one

Refering to the [`lspci`](https://gist.github.com/svmlegacy/734b7b123d7699f34dd2dd5c42876f3f) of two processor sockets populated we have found buses `0x7f` (for sure) but the other one shows as `0x00` where I would expect `0xff` or perhaps `0x80`

Device `0x3c28` where `CPUBUSNO` (of register `0x108`) appears twice:
```
00:05.0 System peripheral [0880]: Intel Corporation Xeon E5/Core i7 Address Map, VTd_Misc, System Management [8086:3c28] (rev 07)

80:05.0 System peripheral [0880]: Intel Corporation Xeon E5/Core i7 Address Map, VTd_Misc, System Management [8086:3c28] (rev 07)
```

We also have another `CPUBUSNO` register to query:

![2022-06-06-084501_775x301_scrot](https://user-images.githubusercontent.com/11563789/172110228-8e56020c-6d11-453c-9f27-cb199a3c30d7.png)

More than that; using E7, according to its [datasheet](https://www.intel.fr/content/dam/www/public/us/en/documents/datasheets/xeon-e7-v2-datasheet-vol-2.pdf), `CPUBUSNO` has to be queried from BDF(1, 11, 3)
![2022-06-06-085819_766x381_scrot](https://user-images.githubusercontent.com/11563789/172112109-62568750-eedf-4107-97be-425cf053f334.png)

**Comment by cyring** [comment]
@svmlegacy Hello,

Please find another archive to dump registers printed as `CoreFreq: @ BDF()  CPUBUSNO[]`

[CoreFreq_develop.tar.gz](https://github.com/cyring/CoreFreq/files/8843252/CoreFreq_develop.tar.gz)

**Reply by svmlegacy** [reply]
```
[  319.217850] CoreFreq(10:22:-1): Processor [ 06_2D] Architecture [SandyBridge/EP/Romley] SMT [24/24]
[  319.218051] CoreFreq: @ BDF(0,5,0) CPUBUSNO[17f00]
```

**Reply by cyring** [reply]
Thank you for this dump.
`BDF(1, 11, 3)` did not work apparently.
It's difficult to find a register value or function result which explains why the IMC devices of **second** socket are listed under bus `0xff` (presuming `0x7f` being first)
Can it be kernel or BIOS choice ?
Can it be reported differently by `lspci` ?

**Reply by cyring** [reply]
> Thank you for this dump. `BDF(1, 11, 3)` did not work apparently. It's difficult to find a register value or function result which explains why the IMC devices of **second** socket are listed under bus `0xff` (presuming `0x7f` being first) Can it be kernel or BIOS choice ? Can it be reported differently by `lspci` ?

Got it; thanks to this `06_3E`
```
CoreFreq: @ BDF(0,5,0) CPUBUSNO[13f00]
CoreFreq: @ BDF(40,5,0) CPUBUSNO[17f40]
```
1. At bus `#0x0` you'll read offset `0x108` for register CPUBUSNO at `BDF(0,5,0)`
  We start at bus `0x0` because we probed Host Bridge at this BDF location
3. Bits [16-8] UNC = `0x3f`
4. Next bus is UNC + 1 = `0x3f + 1` = `0x40`
5. At bus `#0x40` you'll read offset `0x108` for CPUBUSNO at `BDF(0x40,5,0)`
6. Bits [16-8] UNC = `0x7f`

We thus have two memory controllers, one MC at bus `0x3f` and a second at bus `0x7f`

Not sure about the MC order and if `IIO=0x40` matters in the topology ?

EDIT: using Device 11, Function 3, Offset `0xd0` just confirms the mapping
```
CoreFreq: @ BDF(0,5,0) CPUBUSNO[13f00]
CoreFreq: @ BDF(3f,11,3) CPUBUSNO[80003f00]
CoreFreq: @ BDF(40,5,0) CPUBUSNO[17f40]
CoreFreq: @ BDF(7f,11,3) CPUBUSNO[80007f40]
```
Datasheet does not mention an end stop, so I believe 2 IMC is hard coded in architecture

**Comment by cyring** [comment]
@svmlegacy 

Can you please try this version ?
[CoreFreq_develop.tar.gz](https://github.com/cyring/CoreFreq/files/8849699/CoreFreq_develop.tar.gz)

This time, no dump. The Memory Controllers will be decoded following the CPU socket topology.
This test should only show the DRAM primary timings, aka `tCL`, for any IMC, channels possible.

_Ignore the unused functions warnings during compilation_

**Comment by cyring** [comment]
@svmlegacy 
This version will scan all buses for memory controllers
[CoreFreq_develop_xeon_all_mc.tar.gz](https://github.com/cyring/CoreFreq/files/8855950/CoreFreq_develop_xeon_all_mc.tar.gz)

**Reply by svmlegacy** [reply]
Here's a dmesg dump from making and running this package:
https://gist.github.com/svmlegacy/9979f205743c89ab570b2e6fb27881b8

**Reply by svmlegacy** [reply]
![image](https://user-images.githubusercontent.com/59940364/173169385-7a370a91-a074-4414-9fd0-cf62672e6a1e.png)

**Reply by cyring** [reply]
> Here's a dmesg dump from making and running this package: https://gist.github.com/svmlegacy/9979f205743c89ab570b2e6fb27881b8

Since the archive, additional codes should improve the IMC decoding.
Can you pull from the latest `develop` branch and try again ?

Thank you

**Reply by svmlegacy** [reply]
![image](https://user-images.githubusercontent.com/59940364/173197548-f058ee3d-0884-434d-97fe-7478d74e4e67.png)
![image](https://user-images.githubusercontent.com/59940364/173197537-c9fa4c98-6a6b-4af5-85b3-14f430f73cf3.png)

**Reply by svmlegacy** [reply]
It's interesting to see that the memory controller has different values in the unpopulated channels for timings. Everything in the populated channels does look correct at a glace. (And yes, I'm expecting to see channels 1 & 3 on CPU0 and 0 & 2 on CPU1

**Reply by cyring** [reply]
> It's interesting to see that the memory controller has different values in the unpopulated channels for timings. Everything in the populated channels does look correct at a glace. (And yes, I'm expecting to see channels 1 & 3 on CPU0 and 0 & 2 on CPU1

Great progress on the PCI topology. 

QPI rate is also available.

Specs of some third timings are still unknown 

I'm close to merge into Master ...

Thank you.

**Comment by cyring** [comment]
Results from the latest development commit:

![2022-06-09-175138_642x1141_scrot](https://user-images.githubusercontent.com/11563789/172890805-034cbc67-032a-467c-97d9-fbcfc5fb6a1c.png)
